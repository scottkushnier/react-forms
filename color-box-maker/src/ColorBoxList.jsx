
import { useState } from 'react'
import ColorBox from './ColorBox.jsx'
import NewBoxForm from './NewBoxForm.jsx';

let keyInd = 1;
function newKey() {
    const res = keyInd;
    keyInd++;
    return res;
}

// const initialList = [{color: "khaki", height: "30px", width: "30px", key: newKey()},
//   {color: "orange", height: "40px", width: "30px", key: newKey()},
//   {color: "lightgreen", height: "20px", width: "30px", key: newKey()}];

const initialList = [];

function ColorBoxList() {
    const removeBoxWithKey = key => setColorBoxes(colorBoxes.filter(box => (box.key != key)));
    const addBox = (formData) => {
      const {color, height, width} = formData;
      setColorBoxes([...colorBoxes, {color, height: height+'px', width: width+'px', key: newKey()}]);
    }
    const [colorBoxes, setColorBoxes] = useState(initialList);
    return (
        <div data-testid="boxlist">
          {colorBoxes.map(({color, height, width, key}) => {
               return <ColorBox color={color} key={key} height={height} width={width} 
                            deleteFn={() => {removeBoxWithKey(key)}}
                      />
          })}
          <NewBoxForm AddFn={addBox}/>
        </div>
    )
}

export default ColorBoxList;