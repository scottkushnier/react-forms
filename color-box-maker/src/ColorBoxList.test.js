

import { render, fireEvent } from "@testing-library/react";
import ColorBoxList from "./ColorBoxList";
import '@testing-library/jest-dom'

test("colorboxlist smoke test", function () {
  render(<ColorBoxList />)
});

test("matches snapshot", function() {
    const {asFragment} = render(<ColorBoxList />);
    expect(asFragment()).toMatchSnapshot();
  });

test("should add new box", function() {
    const {queryByText, getByLabelText, queryByTestId} = render(<ColorBoxList />);
    const btn = queryByText('Submit');
    const colorInput = getByLabelText('Color:');
    const heightInput = getByLabelText('Height:');
    const widthInput = getByLabelText('Width:');
    const boxList = queryByTestId("boxlist");
    expect(boxList).not.toContainHTML("purple");
    fireEvent.change(colorInput, {target: {value: 'purple'}});
    fireEvent.change(heightInput, {target: {value: 42}});
    fireEvent.change(widthInput, {target: {value: 53}});
    fireEvent.click(btn);
    expect(boxList).toContainHTML("purple");
    expect(boxList).toContainHTML("42px");
    expect(boxList).toContainHTML("53px");
});

