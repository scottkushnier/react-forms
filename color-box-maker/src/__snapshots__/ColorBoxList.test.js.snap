// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="boxlist"
  >
    <form>
      <br />
      New Block
      <br />
      <label
        for="color"
      >
        Color: 
      </label>
      <input
        id="color"
        name="color"
        value="white"
      />
      <br />
      <label
        for="height"
      >
        Height: 
      </label>
      <input
        id="height"
        name="height"
        value="30"
      />
      <br />
      <label
        for="width"
      >
        Width: 
      </label>
      <input
        id="width"
        name="width"
        value="30"
      />
      <br />
      <button>
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
